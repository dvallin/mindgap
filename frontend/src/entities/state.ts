import produce from 'immer'
import { List } from 'lazy-space'

import { BaseAction } from '../react-helpers'

export type EntityCreateAction = BaseAction<'ENTITY_CREATE'> & { entity: Entity }
export const entityCreateAction = (entity: Entity): EntityCreateAction => ({
  type: 'ENTITY_CREATE',
  entity,
})

export type EntityAction = EntityCreateAction

export type Entity = string

export interface EntityState {
  dirty: List<string>
}

export const initialState = (): EntityState => ({ dirty: List.empty() })

export const entity = (state: EntityState = initialState(), action: EntityAction): EntityState => {
  return produce(state, (draft) => {
    switch (action.type) {
      case 'ENTITY_CREATE': {
        draft.dirty = draft.dirty.append(action.entity)
      }
    }
  })
}
