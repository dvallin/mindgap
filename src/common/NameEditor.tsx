import * as React from 'react'

import { Name } from '.'
import EditField from '../ui/EditField'

export interface Props extends Name {
  updateName: (name: Name) => void
}

export class NameEditor extends React.Component<Props> {
  render(): JSX.Element {
    return (
      <>
        <h1 className="title">
          <EditField
            placeholder="Name"
            value={this.props.name}
            onChange={(value) => this.props.updateName({ name: value, description: this.props.description })}
            removePadding
          />
        </h1>
        <h2 className="subtitle">
          <EditField
            placeholder="Description"
            value={this.props.description}
            onChange={(value) => this.props.updateName({ name: this.props.name, description: value })}
            removePadding
          />
        </h2>
      </>
    )
  }
}

export default NameEditor
