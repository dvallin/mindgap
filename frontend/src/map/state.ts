import produce from 'immer'
import { Option, List } from 'lazy-space'

import { BaseAction } from '../react-helpers'
import { Entity } from '../entities/state'

export type PositionUpdatedAction = BaseAction<'POSITION_UPDATED'> & { coordinates: GeoCoordinates; data: string }
export const positionUpdatedAction = (coordinates: GeoCoordinates, data: string): PositionUpdatedAction => ({
  type: 'POSITION_UPDATED',
  coordinates,
  data,
})

export type LocationAddAction = BaseAction<'LOCATION_ADD'> & { entity: Entity; coordinates: GeoCoordinates; data: string }
export const locationAddAction = (entity: Entity, coordinates: GeoCoordinates, data: string): LocationAddAction => ({
  type: 'LOCATION_ADD',
  entity,
  coordinates,
  data,
})

export type LocationDeleteAction = BaseAction<'LOCATION_DELETE'> & { entity: Entity }
export const locationDeleteAction = (entity: Entity): LocationDeleteAction => ({
  type: 'LOCATION_DELETE',
  entity,
})

export type LocationFocusAction = BaseAction<'LOCATION_FOCUS'> & { coordinates: GeoCoordinates }
export const locationFocusAction = (coordinates: GeoCoordinates): LocationFocusAction => ({
  type: 'LOCATION_FOCUS',
  coordinates,
})

export type PositionAction = PositionUpdatedAction | LocationAddAction | LocationDeleteAction | LocationFocusAction

export interface GeoCoordinates {
  latitude: number
  longitude: number
}

export interface Location {
  coordinates: GeoCoordinates
  data: string
}

export interface LocationComponent {
  entity: Entity
  location: Location
}

export interface PositionState {
  current: Option<Location>
  locations: List<LocationComponent>
  focus: Option<GeoCoordinates>
}

export function coordinatesEqual(a: GeoCoordinates, b: GeoCoordinates): boolean {
  return a.latitude === b.latitude && a.longitude === b.longitude
}

export function exactLocation(focus: GeoCoordinates, locations: List<LocationComponent>): Option<LocationComponent> {
  return locations.find((l) => coordinatesEqual(focus, l.location.coordinates))
}

export function nearestLocation(focus: GeoCoordinates, locations: List<LocationComponent>): Option<LocationComponent> {
  let squaredDiff = Number.MAX_VALUE
  let result: Option<LocationComponent> = Option.none()
  locations.map((l) => {
    const latDiff = l.location.coordinates.latitude - focus.latitude
    const lngDiff = l.location.coordinates.longitude - focus.longitude
    const diff = latDiff * latDiff + lngDiff * lngDiff
    if (diff < squaredDiff) {
      squaredDiff = diff
      result = Option.some(l)
    }
  })
  return result
}

export const initialState = (): PositionState => ({ current: Option.none(), locations: List.empty(), focus: Option.none() })

export const position = (state: PositionState = initialState(), action: PositionAction): PositionState => {
  return produce(state, (draft) => {
    switch (action.type) {
      case 'POSITION_UPDATED': {
        draft.current = Option.some({
          coordinates: action.coordinates,
          data: action.data,
        })
        break
      }
      case 'LOCATION_ADD': {
        draft.locations = draft.locations.append({
          entity: action.entity,
          location: {
            coordinates: action.coordinates,
            data: action.data,
          },
        })
        break
      }
      case 'LOCATION_DELETE': {
        draft.locations = draft.locations.filter((v) => v.entity !== action.entity)
        break
      }
      case 'LOCATION_FOCUS': {
        draft.focus = Option.some(action.coordinates)
      }
    }
  })
}
