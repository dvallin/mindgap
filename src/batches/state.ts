import produce from 'immer'
import { Option } from 'lazy-space'

import { Batch } from '.'
import { Id } from '../common'
import { BaseAction } from '../react-helpers'

export interface BatchState {
  batchCache: { [id in Id]: Batch }
}

export type RemoveRecipe = BaseAction<'REMOVE_RECIPE'> & { id: Id }
export const removeRecipe = (id: Id): RemoveRecipe => ({ type: 'REMOVE_RECIPE', id })

export type AddRecipe = BaseAction<'ADD_RECIPE'> & { id: Id; recipe: Id }
export const removedRecipe = (id: Id, recipe: Id): AddRecipe => ({ type: 'ADD_RECIPE', id, recipe })

export type BatchAction = RemoveRecipe | AddRecipe

export const initialState = (): BatchState => ({
  batchCache: {
    batch1: {
      name: 'My Special Kombuch',
      description: 'default kombucha with some chilis',
      recipe: {
        id: 'recipe1',
        scale: 2,
      },
      actions: [],
    },
  },
})

export const batches = (state: BatchState = initialState(), action: BatchAction): BatchState => {
  return produce(state, (draft) => {
    switch (action.type) {
      case 'ADD_RECIPE': {
        Option.of(draft.batchCache[action.id]).map((batch) => (batch.recipe = { id: action.recipe, scale: 1 }))
        break
      }
      case 'REMOVE_RECIPE': {
        Option.of(draft.batchCache[action.id]).map((batch) => delete batch.recipe)
        break
      }
    }
  })
}
