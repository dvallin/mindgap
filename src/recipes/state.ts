import { Recipe } from '.'
import { BaseAction } from '../react-helpers'
import produce from 'immer'
import { Id, Name } from '../common'
import { Ingredient } from '../ingredients'
import { Option } from 'lazy-space'
import { Step } from '../steps'

export interface RecipeState {
  recipeCache: { [id in Id]: Recipe }
}

export type AddRecipe = BaseAction<'ADD_RECIPE'> & { id: string; recipe: Recipe }
export const addRecipe = (id: string, recipe: Recipe): AddRecipe => ({ type: 'ADD_RECIPE', id, recipe })

export type UpdateName = BaseAction<'UPDATE_NAME'> & { id: string; name: Name }
export const updateName = (id: string, name: Name): UpdateName => ({
  type: 'UPDATE_NAME',
  id,
  name,
})
export type AddIngredient = BaseAction<'ADD_INGREDIENT'> & { id: string; ingredient: Ingredient }
export const addIngredient = (id: string, ingredient: Ingredient): AddIngredient => ({ type: 'ADD_INGREDIENT', id, ingredient })
export type DeleteIngredient = BaseAction<'DELETE_INGREDIENT'> & { id: string; index: number }
export const deleteIngredient = (id: string, index: number): DeleteIngredient => ({ type: 'DELETE_INGREDIENT', id, index })
export type UpdateIngredient = BaseAction<'UPDATE_INGREDIENT'> & { id: string; index: number; ingredient: Ingredient }
export const updateIngredient = (id: string, index: number, ingredient: Ingredient): UpdateIngredient => ({
  type: 'UPDATE_INGREDIENT',
  id,
  index,
  ingredient,
})
export type AddStep = BaseAction<'ADD_STEP'> & { id: string; step: Step }
export const addStep = (id: string, step: Step): AddStep => ({ type: 'ADD_STEP', id, step })
export type UpdateStep = BaseAction<'UPDATE_STEP'> & { id: string; index: number; step: Step }
export const updateStep = (id: string, index: number, step: Step): UpdateStep => ({
  type: 'UPDATE_STEP',
  id,
  index,
  step,
})

export type RecipeAction = AddRecipe | UpdateName | AddIngredient | DeleteIngredient | UpdateIngredient | AddStep | UpdateStep

export const initialState = (): RecipeState => ({
  recipeCache: {},
})

export const recipes = (state: RecipeState = initialState(), action: RecipeAction): RecipeState => {
  return produce(state, (draft) => {
    switch (action.type) {
      case 'UPDATE_NAME': {
        Option.of(draft.recipeCache[action.id]).map((recipe) => {
          recipe.description = action.name.description
          recipe.name = action.name.name
        })
        break
      }
      case 'ADD_RECIPE': {
        draft.recipeCache[action.id] = action.recipe
        break
      }
      case 'ADD_INGREDIENT': {
        Option.of(draft.recipeCache[action.id]).map((recipe) => recipe.ingredients.push(action.ingredient))
        break
      }
      case 'DELETE_INGREDIENT': {
        Option.of(draft.recipeCache[action.id]).map((recipe) => recipe.ingredients.splice(action.index, 1))
        break
      }
      case 'UPDATE_INGREDIENT': {
        Option.of(draft.recipeCache[action.id]).map((recipe) => (recipe.ingredients[action.index] = action.ingredient))
        break
      }
      case 'ADD_STEP': {
        Option.of(draft.recipeCache[action.id]).map((recipe) => recipe.steps.push(action.step))
        break
      }
      case 'UPDATE_STEP': {
        Option.of(draft.recipeCache[action.id]).map((recipe) => (recipe.steps[action.index] = action.step))
        break
      }
    }
  })
}
