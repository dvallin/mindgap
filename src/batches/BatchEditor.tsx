import { Option } from 'lazy-space'
import * as React from 'react'
import { connect } from 'react-redux'

import { Batch } from '.'
import { Id, Name } from '../common'

import NameEditor from '../common/NameEditor'
import { Recipe } from '../recipes'
import { Dispatch, State } from '../store'
import { scale as scaleRecipe } from '../recipes'
import { addAction, updateAction, updateName, updateRecipe, updateRecipeScale } from './state'
import produce from 'immer'
import RecipeComponent from '../recipes/Recipe'
import RecipeSelect from '../recipes/RecipeSelect'
import StepsEditor from '../steps/StepsEditor'
import { Step } from '../steps'

export interface Callbacks {
  updateName: (name: Name) => void
  updateRecipe: (id: string) => void
  updateRecipeScale: (scale: number) => void
  addStep: (step: Step) => void
  updateStep: (index: number, step: Step) => void
}

export interface InnerProps {
  batch: Option<Batch>
  recipe: Option<Recipe>
}

export interface OuterProps {
  id: Id
}

export type Props = OuterProps & Callbacks & InnerProps

export const BatchEditor = (props: Props): JSX.Element =>
  props.batch.unwrap(
    (batch) => (
      <div className="tile is-ancestor">
        <div className="tile is-8 is-vertical is-parent">
          <div className="tile is-child">
            <NameEditor {...batch} updateName={props.updateName} />
            <div className="content">
              <StepsEditor steps={batch.actions} addStep={props.addStep} updateStep={props.updateStep} />
            </div>
          </div>
        </div>
        <div className="tile is-parent">
          <div className="tile is-child box">
            <RecipeSelect
              onSelect={props.updateRecipe}
              placeholder="Select a base Recipe"
              value={Option.of(batch.recipe)
                .map((r) => r.id)
                .getOrElse(undefined)}
            />
            <input
              className="input"
              step="1"
              min="0"
              max="100"
              value={Option.of(batch.recipe)
                .map((r) => r.scale)
                .getOrElse(0)}
              onChange={(e) => props.updateRecipeScale(e.target.valueAsNumber)}
              type="range"
            ></input>
            {props.recipe.unwrap(
              (recipe) => (
                <RecipeComponent {...recipe} />
              ),
              () => (
                <></>
              )
            )}
          </div>
        </div>
      </div>
    ),
    () => <></>
  )

export const dispatchToProps = (dispatch: Dispatch, props: OuterProps): Callbacks => ({
  updateName: (name) => dispatch(updateName(props.id, name)),
  updateRecipe: (id) => dispatch(updateRecipe(props.id, id)),
  updateRecipeScale: (scale) => dispatch(updateRecipeScale(props.id, scale)),
  addStep: (step) => dispatch(addAction(props.id, step)),
  updateStep: (index, step) => dispatch(updateAction(props.id, index, step)),
})

export const stateToProps = (state: State, props: OuterProps): InnerProps => {
  const batch = Option.of(state.batches.batchCache[props.id])
  const recipe = batch
    .map((batch) => batch.recipe)
    .flatMap(({ id, scale }) => Option.of(state.recipes.recipeCache[id]).map((r) => produce(r, (d) => scaleRecipe(d, scale))))
  return { batch, recipe }
}

export default connect(stateToProps, dispatchToProps)(BatchEditor)
