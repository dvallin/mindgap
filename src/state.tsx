import { h, createContext, RenderableProps } from 'preact'
import { useContext, useReducer } from 'preact/hooks'
import { Draft, produce } from 'immer'

import { Id } from './common'
import { Batch } from './batches'
import { Recipe } from './recipes'

export type Mutation<State> = (draft: Draft<State>) => void
export type Mutate<State> = (mutation: Mutation<State>) => void

export interface State {
  batches: { [id in Id]: Batch }
  recipes: { [id in Id]: Recipe }
}
export const initialState: State = {
  batches: {},
  recipes: {},
}

export const ApplicationState = createContext<[State, Mutate<State>]>([initialState, () => undefined])

export type Listener = (state: State) => void
export function Provider<T>(props: RenderableProps<T & { listeners: Listener[]; initial: () => State }>) {
  const [state, mutate] = useReducer<State, Mutation<State>>((state, mutation) => {
    const s = produce(state, draft => {
      mutation(draft)
    })
    props.listeners.forEach(l => l(s))
    return s
  }, props.initial())
  return <ApplicationState.Provider value={[state, mutate]}>{props.children}</ApplicationState.Provider>
}

export function useApplicationState(): [State, Mutate<State>] {
  return useContext(ApplicationState)
}
