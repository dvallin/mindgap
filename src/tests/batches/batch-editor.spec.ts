import BatchEditor, { BatchEdit } from '../../batches/batch-editor'

import { initialState, State } from '../../state'

import NotFound from '../../pages/not-found'
import produce from 'immer'
import { createShallowMount, mockApplicationState, testBatch } from '../test-utils'
import NameEditor, { Props as NameProps } from '../../name/name-editor'

describe('BatchEditor', () => {
  const mount = createShallowMount(BatchEditor, { id: 'batch1' })

  it('renders not found', () => {
    const component = mount()
    expect(component).toMatchSnapshot()
    expect(component.exists(NotFound)).toBeTruthy()
  })

  it('renders batch', () => {
    mockApplicationState(
      produce(initialState, s => {
        s.batches['batch1'] = testBatch
      })
    )

    const component = mount()

    expect(component).toMatchSnapshot()
    expect(component.exists(NotFound)).toBeFalsy()
  })
})

describe('BatchEdit', () => {
  const mount = createShallowMount(BatchEdit, { id: 'batch1', batch: testBatch })

  let state: State
  beforeEach(() => {
    mockApplicationState(
      produce(initialState, s => {
        s.batches['batch1'] = testBatch
      }),
      s => (state = s)
    )
  })

  it('renders batch', () => {
    const component = mount()
    expect(component).toMatchSnapshot()
  })

  it('updates names', () => {
    const component = mount()
    component.find<NameProps>(NameEditor).prop('updateName')({ name: 'new name', description: 'new description' })
    expect(state).toMatchObject({ batches: { batch1: { name: 'new name', description: 'new description' } } })
  })
})
