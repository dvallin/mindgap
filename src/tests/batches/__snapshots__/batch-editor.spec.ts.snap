// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchEdit renders batch 1`] = `
Array [
  <Component
    name={
      Object {
        "actions": Array [],
        "date": "2020-10-18:10:00:00Z",
        "description": "",
        "name": "",
      }
    }
    updateName={[Function]}
  />,
  <div
    className="content"
  >
    <Component
      addStep={[Function]}
      deleteStep={[Function]}
      disabled={false}
      steps={Array []}
      updateStep={[Function]}
    />
    <button
      className="button is-info"
      onClick={[Function]}
    >
      <span>
        close
      </span>
    </button>
  </div>,
]
`;

exports[`BatchEditor renders batch 1`] = `
<div
  className="tile is-ancestor"
>
  <div
    className="tile is-8 is-vertical is-parent"
  >
    <div
      className="tile is-child"
    >
      <Component
        batch={
          Object {
            "actions": Array [],
            "date": "2020-10-18:10:00:00Z",
            "description": "",
            "name": "",
          }
        }
        id="batch1"
      />
    </div>
  </div>
  <div
    className="tile is-parent"
  >
    <div
      className="tile is-child box"
    >
      <Component
        batch={
          Object {
            "actions": Array [],
            "date": "2020-10-18:10:00:00Z",
            "description": "",
            "name": "",
          }
        }
        id="batch1"
      />
    </div>
  </div>
</div>
`;

exports[`BatchEditor renders not found 1`] = `<Component />`;
